### PR TITLE
File Manager

### PR DESCRIPTION
**Pull request by mwu-tow**
### Pull Request Description
This PR implements the File Manager — an Akka-based service providing a basic filesystem-like API over network. 

Currently it just wraps the local filesystem operations. In the future support for other kinds of storage might get added.

The code comes with tests suite.

### Important Notes
* the watcher API is significantly limited by the quality of implementation, issues basically stem from the total lack of any guarantees (see https://docs.oracle.com/javase/8/docs/api/java/nio/file/WatchService.html ) — implementations wildly vary in the observed behavior. The notification may or may not come in order, some operations might get missed and so on. 
I struggled quite a while, trying to have possibly extensive tests and have them reliably pass on all three platforms.
* this is my first code written in Scala.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] The code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] The code has been tested where possible.

### CODE REVIEW

File: file-manager/src/main/scala/org/enso/filemanager/FileManager.scala [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+    def touchedPaths: Seq[Path]
+    def handle(fileManager: FileManagerBehavior): ResponseType
+
+    def validate(projectRoot: Path): Unit =
+      touchedPaths.foreach(Detail.validatePath(_, projectRoot))
+  }
+
+  sealed abstract class SuccessResponse
+
+  case class CopyDirectoryRequest(from: Path, to: Path)
```
[the remark this answers is not in the document]
Comment (mwu-tow) [reply file-manager/src/main/scala/org/enso/filemanager/FileManager.scala]: Such separation exists on the OS-provided API level and is typically maintained all the way up.
I find this separation agreeable (or at least not worth hiding), as ordinary files and directories are different enough to usually require user's code to care about what it is dealing with. (e.g. Haskell's `path` package even encodes this on the type-level)

File: file-manager/src/main/scala/org/enso/filemanager/FileManager.scala [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+    }
+  }
+
+  case class ReadResponse(contents: Array[Byte]) extends SuccessResponse
+
+  case class StatRequest(path: Path) extends RequestPayload[StatResponse] {
+    override def touchedPaths: Seq[Path] = Seq(path)
+
+    override def handle(fileManager: FileManagerBehavior): StatResponse = {
+      // warning: might be race'y
```
[the remark this answers is not in the document]
Comment (mwu-tow) [reply file-manager/src/main/scala/org/enso/filemanager/FileManager.scala]: Actually it just should be fixed. When I was writing I wasn't aware that all that attributes can be get in a single call. (well, whether it is actually atomic remains implementation-defined, but we should not make things worse)

**Comment by mwu-tow** [comment]
@iamrecursion 
Thanks for the review!

> In general please don't forget to run scalafmt on everything as well.

Just re-checked it and it doesn't make any additional changes. Are there any changes that should get formatted by scalafmt in a different way? Perhaps I have it misconfigured? Or was it just a general remark and the code is fine?

> our guidelines require doc comments for new code, so I'd like to see some documentation that explains the various constructs in brief.

I will look into adding some doc comments.

Frankly speaking that was not my understanding when I read the guidelines. It says that doc comments are "allowed" but only after a section that makes an impression that they are not really encouraged. Perhaps we should consider changing the guidelines wording so the requirements for the new code are more clearly described.

**Comment by iamrecursion** [comment]
Regarding formatting I swear I saw something that hadn't been formatted, but on a second look it all seems fine. 

Regarding commenting that's my fault if it wasn't clear in the style guides. New code should at least have doc comments to explain usage, and then source notes to explain design elements or reasoning behind why it's written how it is. I'll clarify the points when I'm back at work.

**Pull request merged**